com-api/com-api-concept/src/lib.rs

```diff
@@ -160,48 +239,83 @@ where
     ///
     /// This corresponds to `MaybeUninit::write`.
     fn write(self, value: T) -> Self::SampleMut;
+
+    /// Get a mutable pointer to the internal maybe uninitialized `T`.
+    ///
+    /// The caller has to make sure to initialize the data in the buffer.
+    /// Reading from the received pointer before initialization is undefined behavior.
+    fn as_mut_ptr(&mut self) -> *mut T;
 }
 
-pub trait Interface {}
+pub trait Interface {
+    type Consumer<R: Runtime + ?Sized>: Consumer<R>;
+    type Producer<R: Runtime + ?Sized>: Producer<R>;
+}
 
-pub trait OfferedProducer {
+#[must_use = "if a service is offered it will be unoffered and dropped immediately, causing unexpected behavior in the system"]
+pub trait OfferedProducer<R: Runtime + ?Sized> {
     type Interface: Interface;
-    type Producer: Producer<Interface = Self::Interface>;
+    type Producer: Producer<R, Interface = Self::Interface>;
 
     fn unoffer(self) -> Self::Producer;
 }
 
-pub trait Producer {
+pub trait Producer<R: Runtime + ?Sized> {
     type Interface: Interface;
-    type OfferedProducer: OfferedProducer<Interface = Self::Interface>;
+    type OfferedProducer: OfferedProducer<R, Interface = Self::Interface>;
 
```

Comment:
We will need new here
```suggestion
fn new(instance_info: R::ProviderInfo) -> Self;
```

Reply:
The chain here should be `Runtime::producer_builder().build()`. Would that fit as well?

Reply:
Yes, it was needed to be able to crate Producer with instance_info from ProducerBuilder.